mozilla-services/consvc-shepherd
# [SHEP-17] Add BQ data fetcher script

## References

[SHEP-17](https://mozilla-hub.atlassian.net/browse/SHEP-17)

## Problem Statement

We want to be able to pull some Kevel metadata from BigQuery to display in on the Shepherd UI. This script retrieves selected ad data from BigQuery and inserts it into Shepherd DB. Once configured, this script will run daily, fetching some Kevel metadata for flights.

## Proposed Changes

- **Add the "sync_bq_data" script**: This script will use the BigQuery client to query Kevel metadata in BigQuery. If the data is successfully fetched, it will be inserted into the Shepherd database. If the script runs again, it will update the click and impression data for flights rather than creating new rows in the db.
- **Add unit tests for the sync_bq_data script**: Ensure that the functions in the script are behaving correctly.

## Verification Steps

**Note**: This script includes a field (`flight_name`) that is not yet available in BigQuery. As a result, running the script as is will cause an error because the BigQuery client won't find the flight name. I’m coordinating with Curtis to get this field added, but in the meantime, to test the script, we can comment out or temporarily delete the `flight_name` field from the queries in this script.

To manually run and test the script:

1. Start Shepherd with `docker compose up --build`.
2. Exec into the Shepherd container with `make debug`.
3. Pass your gcloud credentials to the shepherd container to access the BigQuery view `moz-fx-data-shared-prod.ads.consolidated_ad_metrics_daily_pt`:
    1. Login to gcloud locally `gcloud auth application-default login`.
    2. Copy your gcloud creds into the shepherd container:
   ```bash
   docker cp ~/.config/gcloud/application_default_credentials.json consvc-shepherd-app-1:/app/application_default_credentials.json
   
   export GOOGLE_APPLICATION_CREDENTIALS="/app/application_default_credentials.json"

   ```
4. Run the script. Note that date is optional and defaults to today's date.
    ```bash
    python manage.py sync_bq_data --project_id "moz-fx-ads-nonprod" --date $(date +%F)
    ```
    We are using the `moz-fx-ads-nonprod` project ID to test this.
    
    If successful you should see: `BigQuery fetch completed for data <today's date>`.
5. Now the `consvc_shepherd_deliveredflight` db table was populated with click and impression data. You can verify this by checking the table in PostgreSQL (exec into `db-1` container), or by viewing the Delivered flights tab in the admin interface

6. Last but no least, run the tests and ensure they pass.


[SHEP-17]: https://mozilla-hub.atlassian.net/browse/SHEP-17?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
